Helm: refactoring and optimization

BogdanScherban opened the pull request:
1) Main theme color is defined in index.html. Remove constant MAIN_THEME_COLOR from code generally
2) Add checking, that FeedsPanels and FeedsSelectors aren't null
3) Background image and Card background image are defined in version/ path. If these images are absent, we use theme main color instead of image.

Why I did it?
Because it should work, if we remove all non-core plugins from version/

I found these problems when worked on Scotland-version.

BogdanScherban commented:
I started work with Scotland, and removed all non-core elements from version/ path.

After this I found some problem in Core/ part. So, these changes - to solve these problems.
If it is necessary, I can explain each change

fzaninotto commented on src/core/config/styles.js:
```diff
@@ -1,22 +1,37 @@
+ */
+export function getCurrentThemeColor(isContrastMode = false) {
+    let currentThemeColor = (window && window.config) ? window.config.mainColor : DEFAULT_COLOR;
+    return (isContrastMode) ? "#000" : currentThemeColor;
```
I think the config should be able to specify the contrast mode color as well. So I'd write something like:

```js
 return (isContrastMode)
   ? (window && window.config) ? window.config.mainContrastColor : CONTRAST_COLOR
   : (window && window.config) ? window.config.mainColor : DEFAULT_COLOR;
```

But material ui has a support for themes as well, so I advise you to use those. It's well thought and IMHO offers everything you need in terms of customization.

https://material-ui.com/customization/themes/#type-light-dark-theme

BogdanScherban replied:
Can I used the first solution temporary?
Could we provide Theme via Material UI later? Because right now we should start work with Scotland version - this issue has more priority.

BogdanScherban replied:
Done

fzaninotto commented on src/core/config/styles.js:
```diff
@@ -1,22 +1,37 @@
+    const cardBackgroundImage = get(themeImages, 'cardBackgroundImage', null);
+    let result = (window && window.config) ? window.config.mainColor : DEFAULT_COLOR;
+    if (cardBackgroundImage) {
+        result = 'url(' + cardBackgroundImage + ') 0 0 repeat';
```
use string interpolation

BogdanScherban replied:
Done

fzaninotto commented on src/core/config/styles.js:
```diff
@@ -27,6 +42,8 @@ function getCurrentThemeColor(isContrastMode) {
 
 export function getCurrentTheme(isContrastMode) {
```
You can export one theme with both light and dark variants, so you don't need to choose the mode when calling the function. See https://material-ui.com/customization/themes/#type-light-dark-theme

BogdanScherban replied:
It will be done in future

BogdanScherban replied:
Done

fzaninotto commented on src/version/config/theme.config.js:
```diff
@@ -1,5 +1,8 @@
 import { connect } from 'react-redux';
```
I don't understand the location of this file. Any theme configuration should be either in the core, or in one of the apps, but I don't understand what this one configures.

BogdanScherban replied:
Removed

fzaninotto commented on src/version/features/UserTour/content.js:
```diff
@@ -1,5 +1,9 @@
 
+import { getCurrentThemeColor } from "../../../core/config/styles";
+
+const themeColor = getCurrentThemeColor();
```
no, use a material-ui style in the function form

BogdanScherban replied:
It will be done in future

BogdanScherban replied:
Done

fzaninotto commented on src/version/features/UserTour/content.js:
```diff
@@ -81,7 +85,7 @@ const lastStep = {
         <div className="tour-body">
             <p className="tour-body-content">
-                For more information and a guide on how to use {homepage.title}, please go to <a className="link" href={homepage.link} title={homepage.title} style={{color:"#3596f4", 'text-decoration': "none" }} target="_blank">
+                For more information and a guide on how to use {homepage.title}, please go to <a className="link" href={homepage.link} title={homepage.title} style={{color: themeColor, 'text-decoration': "none" }} target="_blank">
```
use material-ui's CSS in JS instead of styles

BogdanScherban replied:
Done

fzaninotto commented on src/version/features/UserTour/styles.js:
```diff
@@ -1,4 +1,6 @@
-import { MAIN_THEME_COLOR } from "../../config/theme.config";
+import { getCurrentThemeColor } from "../../../core/config/styles";
+
+const themeColor = getCurrentThemeColor();
```
Again, I don't understand why you don't export a mui theme instead

BogdanScherban replied:
Done

BogdanScherban commented:
@fzaninotto 
Francois, I fixed all notes, besides contrast mode via material-ui theme. Can we do this later?

I need in time to understand, which styles settings should we create (because current styles are more flexible in comparisom with initial material-ui styles)

fzaninotto commented:
as for the styles, it's up to @tony-shannon to prioritize.

fzaninotto commented:
Also, you can put custom elements in a material ui theme, to theme your own elements.

BogdanScherban commented:
@fzaninotto 
Francois, I changed color defenition via Material-UI theme settings. I created two stylesheets - for light and for dark mode, and removed all redundant styles.

fzaninotto commented on package.json:
```diff
@@ -12,7 +12,7 @@
     "jsonwebtoken": "^8.4.0",
     "lodash": "^4.17.11",
-    "material-ui": "^0.20.2",
+    "material-ui": "^1.0.0-beta.47",
```
Oh, I hadn't noticed this one... Your dependency is badly outdated for this version. Upgrading to a more recent version of material-ui will require some work, you'd better not delay that task too much.

BogdanScherban replied:
Updated

fzaninotto commented on public/index.html:
```diff
@@ -42,6 +42,9 @@
 	<script type="text/javascript">
         window.config = {
             domainName: "https://192.168.88.38",
+            mainColor: "#3596f4",
```
in my opinion, you should just store one `theme` object in the config, and merge this with the default theme when calling `createTheme()`. You should not use the color settings from window anywhere else than in the theme.

BogdanScherban replied:
Done

BogdanScherban commented:
@fzaninotto 
What I did:
1) Update material-ui to the latest version;
2) Fix all styles, which were broken because of it;
3) Create **theme** object in **window.config** and merge it with default theme in createTheme()

PR was updated.

fzaninotto commented on package.json:
```diff
@@ -8,11 +8,11 @@
     "@fortawesome/fontawesome-svg-core": "^1.2.12",
     "@fortawesome/free-solid-svg-icons": "^5.6.3",
     "@fortawesome/react-fontawesome": "^0.1.4",
+    "@material-ui/core": "^3.9.2",
```
react-admin is only compatible with react-admin 1.5, you're going too far ;)

fzaninotto commented on public/index.html:
```diff
@@ -42,9 +42,11 @@
-            mainColor: "#3596f4",
-            dangerColor: "#da534f",
-            contrastColor: "#000",
+			theme: {
```
indentation issue?

BogdanScherban replied:
Solved - two palettes were added to window.config.

fzaninotto commented on src/core/config/styles.js:
```diff
@@ -32,23 +35,21 @@ function getCardBackground(isContrastMode = false) {
     const backgroundImage = get(themeImages, 'backgroundImage', null);
-    const cardBackground = getCardBackground(isContrastMode);
-
+    const themeColor = (window && window.config) ? Object.assign({}, defaultTheme, window.config.theme) : defaultTheme;
```
Why do you create an intermediate object? Merge the theme from config with the theme you're building in the following lines. That way, the theme from config can override all the settings you define in this theme. Something like:

```js
const theme = createMuiTheme(deepMerge(defaultTheme, window.config.theme));
``` 

That means the theme from the window.config will have to use the same structure as the mui theme, and store the colors under the palette key.

BogdanScherban replied:
@fzaninotto 
If I understand you correctly, I should do smth like this:
```
// index.html
window.config = {
            domainName: "https://192.168.88.38",
	    palette: {
                  mainColor: "#3596f4",
                  dangerColor: "#da534f",
                  contrastColor: "#000",
	     },
        }

// src/core/config/styles.js
const defaultTheme = {
     palette: {},
     ....
}

return createMuiTheme(deepmerge(defaultTheme,  window.config));
```

I tried to do this, but in this case we have three problems:

1) We use two types of palettes depends on isContrastMode. If we will use deepmerge, window.config will override this one:
```
palette: isContrastMode ? darkPalette : lightPalette,
```

2) We need in theme main color for some places, where we use image as background. If image is absent, we use main theme color:

```
const cardBackground = getCardBackground(isContrastMode, themeColor);
return createMuiTheme({
    tableHeader: {
            tableHeaderBlock: {
                background: cardBackground,
            },
        },
});
```
Or we should add this parameter to window.config.

3) We need in theme main color from window.config in UserTour component (I described this issue below).

fzaninotto replied:
1. Store two palettes in the config

```js
window.config = {
            domainName: "https://192.168.88.38",
	    lightPalette: {
                  mainColor: "#3596f4",
                  dangerColor: "#da534f",
                  contrastColor: "#000",
	     },
            darkPalette: {
                  ...
            }
        }
```

2. I'm not sure I understand your question. If you need the theme color, use the function form of withStyles()

```js
const styles = theme =>> ({
    tableHeader: {
        background: theme.palette.main
    }
});
```

3. I'll answer below

BogdanScherban replied:
Added both palettes to window.config. They are merged with default theme in createMuiTheme: 
https://github.com/PulseTile/PulseTile-RA/blob/feature-optimization/src/core/config/styles.js

fzaninotto commented on src/version/features/UserTour/styles.js:
```diff
@@ -1,7 +1,5 @@
-
-const themeColor = (window && window.config) ? window.config.mainColor : DEFAULT_MAIN_COLOR;
-const dangerColor = (window && window.config) ? window.config.dangerColor : DEFAULT_DANGER_COLOR;
+const themeColor = (window && window.config) ? window.config.theme.mainColor : "#0D672F";
```
again, you should not access the window.config.theme here. It should only be accessed when you create the mui theme.

BogdanScherban replied:
@fzaninotto 

This file includes styles for Joyride-component, which is used for User Tour functionality:
```
// UserTour/styles.js
export default {
    options: {
         ....
    },
    tooltipTitle: {
         ....
    },
}

//UserTour/fragments/RunTourButton.js
import Joyride from 'react-joyride';
import toursStyles from "../styles";

            <Joyride
                ....
                styles={toursStyles}
            />
```

I tried to use Material-UI function withStyles() here and set styles via something like:
```
// UserTour/styles.js
export default theme => ({
    options: {
         ....
    },
    tooltipTitle: {
         ....
    },

});
```
But this variant doesn't work. For this reason can't use theme settings here.

fzaninotto replied:
USe a custom component syles with withStyles() and inject it to Joyride, see https://docs.react-joyride.com/custom-components

BogdanScherban replied:
Unfortunately, not all components can be override by this way (look Props Inheritance):
https://docs.react-joyride.com/step

We can't override buttons, where we want to change the color by the theme like 
```
// ButtonNext.js
const styles = theme => ({});

const ButtonNext = ({ classes, ...props }) => {
     // our button
};

export default withStyles(styles)(ButtonNext);

 // RunTourButton.js
const RunTourButton = ({ classes, ...rest }) => {
    return (
       <Joyride
         ....
         buttonNextComponent={<ButtonNext {...props} />}
         ....
       />
    );
};
```

We can do following:
```
const styles = {
     ...
     ...
};

const JoyRideWithStyles = ({ classes, ...rest }) => {
    return (
              <Joyride
                continuous
                disableOverlayClose={true}
                showSkipButton={true}
                showProgress={true}
                locale={locale}
                steps={tourSteps}
                run={isPassed}
                styles={{classes}}
                callback={callback}
            />
    );
}

export default withStyles(styles)(JoyRideWithStyles);
```

But it will not work. Because withStyles() returns object with class names. Joyride doesn't use class names of its elements (buttons, tooltips etc.):
![usertour](https://user-images.githubusercontent.com/39263134/53819042-d5109980-3f71-11e9-969a-eec5958a06eb.png)

BogdanScherban commented:
@fzaninotto 

Francois, I changed version of Material-UI to 1.5.0 and updated the PR.

I tried to solve the issue with overriding default theme by window.config. But in this case we loose possibility to toggle contrast mode and to set background for cards, if card image is absent.

Besides this, we should use window.config information in UserTour component, because react-joyride library doesn't work with Material-UI
